baybe/utils/augmentation.py

```diff
@@ -0,0 +1,210 @@
+        | a  | b  | x  | y  |
+        +----+----+----+----+
+        | b  | a  | x  | z  |
+        +----+----+----+----+
+        | b  | a  | y  | x  |
+        +----+----+----+----+
+        | a  | b  | z  | x  |
+        +----+----+----+----+
+
+    Args:
+        df: The dataframe that should be augmented.
+        column_groups: Sequences of permutation invariant columns. The n'th column in
+            each group will be permuted together with each n'th column in the other
+            groups.
+
+    Returns:
+        The augmented dataframe containing the original one. Augmented row indices are
+        identical with the index of their original row.
+
+    Raises:
+        ValueError: If less than two column groups are given.
+        ValueError: If a column group is empty.
+        ValueError: If the column groups have differing amounts of entries.
+    """
+    # Validation
+    if len(column_groups) < 2:
+        raise ValueError(
+            "When augmenting permutation invariance, at least two column sequences "
+            "must be given."
+        )
+
+    if len({len(seq) for seq in column_groups}) != 1:
+        raise ValueError(
+            "Permutation augmentation can only work if the amount of columns in each "
+            "sequence is the same."
+        )
+    elif len(column_groups[0]) < 1:
+        raise ValueError(
+            "Permutation augmentation can only work if each column group has at "
+            "least one entry."
+        )
+
+    # Augmentation Loop
+    new_rows: list[pd.DataFrame] = []
+    idx_permutation = list(permutations(range(len(column_groups))))
+    for _, row in df.iterrows():
+        # For each row in the original df, collect all its permutations
+        to_add = []
+        for perm in idx_permutation:
+            new_row = row.copy()
+
+            # Permute columns, this is done separately for each tuple of columns that
+            # belong together
+            for deps in map(list, zip(*column_groups)):
+                new_row[deps] = row[[deps[k] for k in perm]]
+
+            to_add.append(new_row)
+
+        # Store augmented rows, keeping the index of their original row
+        new_rows.append(
+            pd.DataFrame(to_add, columns=df.columns, index=[row.name] * len(to_add))
+        )
+
+    return pd.concat(new_rows)
+
+
+def df_apply_dependency_augmentation(
+    df: pd.DataFrame,
+    causing: tuple[str, Sequence],
+    affected: Iterable[tuple[str, Sequence]],
+) -> pd.DataFrame:
+    """Augment a dataframe if dependency invariant columns are present.
+
+    This works with the concept of column-values pairs for causing and affected column.
+    Any row present where the specified causing column has one of the provided values
+    will trigger an augmentation on the affected columns. The latter are augmented by
+    going through all their invariant values and adding respective new rows.
+
+    *   Original
+
+        +---+---+---+---+
```

Comment:
its an annoying doctest because it requires specifying and printing a complex object so Im not sure

Reply:
Perhaps it *could* be done rather easily by resorting to some of the more high-level print methods of pandas, such as `to_markdown`, which also create tabular-like output for you. But I'm also not sure if it's the best idea. In particular, we should ensure that the tables are still nicely rendered in the docs, which I think is a big plus. But having a doctest is generally a good idea, since you never know if the output will change at some point, e.g. even now while we are still working on the PR

Perhaps you can give it a 5min try and see if it is easily doable or not? Would appreciate

Reply:
Doe sthis work for you?
```python
    Examples:
        >>> df = pd.DataFrame({'A':[1,2],'B':[3,3], "C": [5, 4]})
        >>> dfa = df_apply_permutation_augmentation(df, [['A'], ['B']])
        >>> dfa.to_dict(orient='list')
        {'A': [1, 3, 2, 3], 'B': [3, 1, 3, 2], 'C': [5, 5, 4, 4]}
```

should this replace the tables I wrote in the docstring above or do we just need one example?

Reply:
that what it looks like in the doc
![image](https://github.com/emdgroup/baybe/assets/17951239/c7368c30-2886-4587-984a-24656e84d37d)

Reply:
Hm, clearly not as nice as your original tables. So I'm currently drawn back and forth between nice doc and well-tested code. Probably tendency toward the latter, because it's more an internal functionality than a user-facing one. However, in the ideal world, the code output would simply get rendered nicely in the docs. What if you use `to_markdown` instead, does it make it any better?

Reply:
tried that loks like this 
![image](https://github.com/emdgroup/baybe/assets/17951239/af9e3a59-94d4-4342-b1db-97f5469a361f)

Reply:
hmm what about simply this:
![image](https://github.com/emdgroup/baybe/assets/17951239/74211041-eb1e-4df9-bec4-2b20c93ea9ac)

![image](https://github.com/emdgroup/baybe/assets/17951239/d09c0cbd-08f0-4166-bb43-be4f41caa64f)

Reply:
I think that's sufficiently nice, no? You could even show the original once at the top, then we have pretty much 1:1 the content of the docstring. We still lose the nice rendering, but I can accept that for the advantage of having stuff properly tested. What do you think?

Reply:
Just make sure the tests are actually triggered (e.g. by modifying a number and seeing if it raises the error)

Reply:
already did, it seems toe xecute the entire block and points out the line where it fails if theres a problem

i removed the origninal comment tables as theyre not necessary anymore imo